### PR TITLE
reader: dictionary: disable fuzzy searching for CJK words

### PR DESCRIPTION
**Pull request by cyphar**
Fuzzy searching doesn't work with CJK text because sdcv doesn't handle
CJK text particularly well when fuzzy searching. With Japanese, fuzzy
searching results in large numbers of useless results because sdcv
decides to strip off the wrong part of the word when fuzzy searching.

For a specific example -- 喋られる is "fuzzy searched" by removing the
喋 kanji and searching for られる with some other fuzzy factors, which
results in hundreds of results. However, none of these results are 喋る
(the correct result) because the first step removed the only part of the
word being searched that is relevant. With the Japanese plugin, KOReader
can correctly auto-select and lookup any gramatically correct Japanese
word, meaning that fuzzy searching is not only unnecessary but also
detremental (for any failed lookup candidate in the deconjugation list
we provide, sdcv will do a fuzzy search).

Unfortuantely it's difficult to identify if a word is definitely
Japanese and not Korean or Chinese. However, since it seems unlikely
that sdcv correctly handles Korean or Chinese, it's simpler to just
disable fuzzy searching on all CJK-containing word lookups.

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8458)

### CODE REVIEW

**Comment by poire-z** [comment]
You really think it's something that has to be handled in "rawSdcv" ? I trust you if you do.

Thought it would feel better around:
https://github.com/koreader/koreader/blob/0824886fde58cc7a627e8ab2f4ac647261aaf6e2/frontend/apps/reader/modules/readerdictionary.lua#L801-L814
Like having line 808:
`local candidates, disable_fuzzy_search = self.ui.languagesupport:extraDictionaryFormCandidates(word)`
and the language support plugin would decide depending on what it managed to do.

**Comment by Frenzie** [comment]
I agree with @poire-z that it doesn't feel very raw. But if it's awkward otherwise, that's okay.

**Comment by cyphar** [comment]
Good point, fixed.

**Review by Frenzie — APPROVED** [review]

File: frontend/apps/reader/modules/readerdictionary.lua [hunk 1]
```diff
@@ -811,6 +811,22 @@ function ReaderDictionary:startSdcv(word, dict_names, fuzzy_search)
...
     end
 
+    -- If every word contains a CJK character, every word candidate is
+    -- (probably) a CJK word. We don't want fuzzy searching in this case
+    -- because sdcv cannot handle CJK text properly when fuzzy searching (with
+    -- Japanese, it returns hundreds of useless results).
+    local shouldnt_fuzzy_search = true
+    for _, word in ipairs(words) do
+        if not util.hasCJKChar(word) then
+            shouldnt_fuzzy_search = false
```
Comment (yparitcher) [inline frontend/apps/reader/modules/readerdictionary.lua:821]:
```lua
    local shouldnt_fuzzy_search = false
    for _, word in ipairs(words) do
        if util.hasCJKChar(word) then
            shouldnt_fuzzy_search = true
```
Would have been simpler to read

Comment (cyphar) [reply frontend/apps/reader/modules/readerdictionary.lua:821]: That is not the same logic. This PR only disables fuzzy searching if *every* word in the array has CJK characters. Your suggestion would disable fuzzy searching if *any* word has CJK characters. In practice they might always give the same results, but it's not the same logic.

**Pull request merged**